libs/cangen/src/can_gen_decode.rs

```diff
@@ -0,0 +1,213 @@
+                    Some(true) => {
+                        let mut topic_suffix_point = self.points.remove(0);
+                        let topic_suffix_read = topic_suffix_point.gen_decoder_fn();
+                        let topic = format!("{}/", self.name);
+                        let point_decoders = self
+                            .points
+                            .iter_mut()
+                            .map(|point| {
+                                let decoder = point.gen_decoder_fn();
+                                quote! { decoded_points.push(#decoder); }
+                            })
+                            .collect::<Vec<ProcMacro2TokenStream>>()
+                            .into_iter()
+                            .fold(ProcMacro2TokenStream::new(), |mut acc, ts| {
+                                acc.extend(ts);
+                                acc
+                            });
+                        quote! {
+                            topic_suffixes.push(#topic_suffix_read);
+                            #point_decoders
+                            result.push(
+                                DecodeData::new(decoded_points.clone(),
+                                    #topic + String::from(topic_suffixes.pop()),
+                                    #unit)
+                            );
+                            decoded_points.clear();
+                        }
+                    }
+                    _ => {
+                        let point_decoders = self
+                            .points
+                            .iter_mut()
+                            .map(|point| {
+                                let decoder = point.gen_decoder_fn();
+                                quote! { decoded_points.push(#decoder); }
+                            })
+                            .collect::<Vec<ProcMacro2TokenStream>>()
+                            .into_iter()
+                            .fold(ProcMacro2TokenStream::new(), |mut acc, ts| {
+                                acc.extend(ts);
+                                acc
+                            });
+                        let topic = self.name.clone();
+                        quote! {
+                            #point_decoders
+                            result.push(
+                                DecodeData::new(decoded_points.clone(),
+                                    #topic,
+                                    #unit)
+                            );
+                            decoded_points.clear();
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn gen_decoder_skip(&mut self) -> ProcMacro2TokenStream {
+        quote! { }
+    }
+
+    fn gen_decoder_map_entry(&mut self) -> ProcMacro2TokenStream {
+        quote! { }
+    }
+}
+
+/**
+ *  Function to generate formatted reader line for decoding a CANPoint
+ */
+impl CANGenDecode for CANPoint {
```

Comment:
ya the whole impl thing kinda confuses me ngl.  may need u to explain in person Sunday.

Reply:
[Same as above response](https://github.com/Northeastern-Electric-Racing/Calypso/pull/67#discussion_r1826861840)